hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java

```diff
@@ -967,6 +974,96 @@ public static HoodieData<HoodieRecord> convertFilesToColumnStatsRecords(HoodieEn
+    });
+    allRecordsRDD = allRecordsRDD.union(deletedFilesRecordsRDD);
+
+    List<Pair<String, Map<String, Long>>> partitionToAppendedFilesList = partitionToAppendedFiles.entrySet()
+        .stream().map(entry -> Pair.of(entry.getKey(), entry.getValue())).collect(Collectors.toList());
+    List<Pair<String, String>> filesToAppended = partitionToAppendedFilesList.stream().flatMap(e -> e.getRight().keySet().stream().map(f -> Pair.of(e.getLeft(), f))).collect(Collectors.toList());
+    parallelism = Math.max(Math.min(filesToAppended.size(), recordsGenerationParams.getRecordLevelIndexParallelism()), 1);
+    HoodieData<Pair<String, String>> partitionToAppendedFilesRDD = engineContext.parallelize(filesToAppended, parallelism);
+
+    HoodieData<HoodieRecord> appendedFilesRecordsRDD = partitionToAppendedFilesRDD.filter(p -> FSUtils.isBaseFile(new Path(p.getRight()))).flatMap(partitionToDeletedFilePair -> {
+      return getRecordIndexFromParquetFile(new Configuration(), datasetBasePath, partitionToDeletedFilePair, false, keyGeneratorOpt);
+    });
+
+    allRecordsRDD = allRecordsRDD.union(appendedFilesRecordsRDD);
+
+    return allRecordsRDD;
+  }
+
+  private static Iterator<HoodieRecord> getRecordIndexFromParquetFile(
+      Configuration conf,
+      String datasetBasePath,
+      Pair<String, String> partitionToFilePair,
+      boolean isDeleted,
+      Option<BaseKeyGenerator> keyGeneratorOpt) {
+    final String partition = getPartitionIdentifier(partitionToFilePair.getLeft());
+    final String fileName = partitionToFilePair.getRight();
+    final String fileId = FSUtils.getFileId(fileName);
+    final String fileCommitTime = FSUtils.getCommitTime(fileName);
+
+    Path dataFilePath = new Path(datasetBasePath, String.format("%s%s%s", partition, Path.SEPARATOR, fileName));
+
+    BaseFileUtils baseFileUtils = BaseFileUtils.getInstance(dataFilePath.toString());
+    Iterator<HoodieKey> recordKeyIterator;
+    if (keyGeneratorOpt.isPresent()) {
+      recordKeyIterator = baseFileUtils.getHoodieKeyIterator(conf, dataFilePath, keyGeneratorOpt);
+    } else {
+      recordKeyIterator = baseFileUtils.getHoodieKeyIterator(conf, dataFilePath);
+    }
+    // final List<Long> blockRecordSize = baseFileUtils.getBlockRecordSize(conf, dataFilePath);
+
+    return new Iterator<HoodieRecord>() {
+      long size = 0L;
+
+      @Override
+      public boolean hasNext() {
+        return recordKeyIterator.hasNext();
+      }
+
+      @Override
+      public HoodieRecord next() {
+        size++;
+        // int rowGroupIndex = findRowGroupIndex();
+        int rowGroupIndex = -1;
+        HoodieKey next = recordKeyIterator.next();
+        return HoodieMetadataPayload.createRecordLevelIndexRecord(next.getRecordKey(), next.getPartitionPath(), fileId, rowGroupIndex, isDeleted, fileCommitTime, HoodieOperation.INSERT);
+      }
+
+      // int findRowGroupIndex() {
```

Comment:
guess its not yet used as of this patch. but w/ hfile, may I know how does this row group index help us?

Reply:
RowGroupIndex a reserved attribute, which indicates the serial number of the rowGroup in the parquet file, and it has not been collected in this pr.